corehq/motech/tests/test_value_source.py

```diff
@@ -278,6 +281,123 @@ def test_one_value(self):
+        self.assertTrue(can_import_and_export)
+
+    def test_check_default(self):
+        value_source = get_constant_spam()
+        can_import = value_source.check_direction(DIRECTION_IMPORT)
+        can_export = value_source.check_direction(DIRECTION_EXPORT)
+        can_import_and_export = value_source.check_direction(DIRECTION_BOTH)
+        self.assertTrue(can_import)
+        self.assertTrue(can_export)
+        self.assertTrue(can_import_and_export)
+
+
+class CheckSerializeDirectionTests(SimpleTestCase):
+
+    def test_serialize_in(self):
+        value_source = get_constant_spam({"direction": "in"})
+        self.assertIsNone(value_source.serialize("spam"))
+
+    def test_serialize_out(self):
+        value_source = get_constant_spam({"direction": "out"})
+        self.assertEqual(value_source.serialize("spam"), "spam")
+
+    def test_serialize_both(self):
+        value_source = get_constant_spam({"direction": None})
+        self.assertEqual(value_source.serialize("spam"), "spam")
+
+    def test_serialize_default(self):
+        value_source = get_constant_spam()
+        self.assertEqual(value_source.serialize("spam"), "spam")
+
+
+class CheckDeserializeDirectionTests(SimpleTestCase):
+
+    def test_deserialize_in(self):
+        value_source = get_constant_spam({"direction": "in"})
+        self.assertEqual(value_source.deserialize("spam"), "spam")
+
+    def test_deserialize_out(self):
+        value_source = get_constant_spam({"direction": "out"})
+        self.assertIsNone(value_source.deserialize("spam"))
+
+    def test_deserialize_both(self):
+        value_source = get_constant_spam({"direction": None})
+        self.assertEqual(value_source.deserialize("spam"), "spam")
+
+    def test_deserialize_default(self):
+        value_source = get_constant_spam()
+        self.assertEqual(value_source.deserialize("spam"), "spam")
+
+
+class GetValueDirectionTests(SimpleTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.info = CaseTriggerInfo(
+            domain="test-domain",
+            case_id="123456",
+        )
+
+    def test_get_value_in(self):
+        value_source = get_constant_spam({"direction": "in"})
+        self.assertIsNone(value_source.get_value(self.info))
+
+    def test_get_value_out(self):
+        value_source = get_constant_spam({"direction": "out"})
+        value = value_source.get_value(self.info)
+        self.assertEqual(value, "spam")
+
+    def test_get_value_both(self):
+        value_source = get_constant_spam({"direction": None})
+        self.assertEqual(value_source.get_value(self.info), "spam")
+
+    def test_get_value_default(self):
+        value_source = get_constant_spam()
+        self.assertEqual(value_source.get_value(self.info), "spam")
+
+
 def test_doctests():
     results = doctest.testmod(corehq.motech.value_source)
     assert results.failed == 0
+
+
+def get_constant_spam(update: dict = None) -> ValueSource:
```

Comment:
fancy. I like this syntax

Reply:
Yeah, I find it's especially useful when you're refactoring, because the IDE (if it supports it) highlights everywhere it thinks you're making a mistake. ... But Python doesn't enforce strong typing at runtime, so you can still use duck-typing. Best of both worlds, imho.